### PR TITLE
Add foreach_key_range method to btree iterator, to scan a range of

### PR DESCRIPTION
**Pull request by toregge**
the tree and call function for each key.

@geirst : please review

### CODE REVIEW

File: vespalib/src/vespa/vespalib/btree/btreeiterator.h [hunk 1]
```diff
@@ -303,6 +303,38 @@ class BTreeIteratorBase
      * @param pathSize     New tree height (number of levels of internal nodes)
      */
     VESPA_DLL_LOCAL void clearPath(uint32_t pathSize);
+
+    template <typename FunctionType>
+    void
+    foreach_key_range_start(uint32_t level, FunctionType func) const
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/btree/btreeiterator.h:309]: Please add doc describing function.

File: vespalib/src/vespa/vespalib/btree/btreeiterator.h [hunk 2]
```diff
@@ -303,6 +303,38 @@ class BTreeIteratorBase
...
+            uint32_t idx = _path[level].getIdx();
+            node->foreach_key_range(store, idx + 1, node->validSlots(), func);
+        } else {
+            _leaf.getNode()->foreach_key_range(_leaf.getIdx(), _leaf.getNode()->validSlots(), func);
+        }
+    }
+
+    template <typename FunctionType>
+    void
+    foreach_key_range_end(uint32_t level, FunctionType func) const
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/btree/btreeiterator.h:325]: Please add doc describing function.

File: vespalib/src/vespa/vespalib/btree/btreeiterator.h [hunk 3]
```diff
@@ -451,6 +483,54 @@ class BTreeIteratorBase
...
+            foreach_key_range_start(_pathSize, func);
+            return;
+        }
+        assert(_pathSize == end_itr._pathSize);
+        assert(_allocator == end_itr._allocator);
+        uint32_t level = _pathSize;
+        if (level > 0u) {
+            uint32_t idx;
+            uint32_t eidx;
+            do {
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/btree/btreeiterator.h:504]: Please add a comment on what this the result of this while loop.

File: vespalib/src/vespa/vespalib/btree/btreeiterator.h [hunk 4]
```diff
@@ -451,6 +483,54 @@ class BTreeIteratorBase
...
+                if (idx > eidx) {
+                    return;
+                }
+                if (idx != eidx) {
+                    ++level;
+                    break;
+                }
+            } while (level != 0);
+            if (level > 0u) {
+                foreach_key_range_start(level - 1, func);
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/btree/btreeiterator.h:518]: Please add comment describing the 3 steps done here (left, middle, and right sub trees).

File: vespalib/src/vespa/vespalib/btree/btreeiterator.h [hunk 5]
```diff
@@ -451,6 +483,54 @@ class BTreeIteratorBase
             _leafRoot->foreach_key(func);
         }
     }
+
+    template <typename FunctionType>
+    void
+    foreach_key_range(const BTreeIteratorBase &end_itr, FunctionType func) const
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/btree/btreeiterator.h:489]: Please add doc describing function.

File: vespalib/src/vespa/vespalib/btree/btreenode.h [hunk 6]
```diff
@@ -459,6 +475,16 @@ class BTreeLeafNode : public BTreeNodeTT<KeyT, DataT, AggrT, NumSlots>
         }
     }
 
+    template <typename FunctionType>
+    void foreach_key_range(uint32_t start_idx, uint32_t end_idx, FunctionType func) const {
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/btree/btreenode.h:479]: Please add doc describing function.

File: vespalib/src/vespa/vespalib/btree/btreenode.h [hunk 7]
```diff
@@ -370,6 +370,22 @@ class BTreeInternalNode : public BTreeNodeTT<KeyT, BTreeNode::Ref, AggrT,
         }
     }
 
+    template <typename NodeStoreType, typename FunctionType>
+    void foreach_key_range(NodeStoreType &store, uint32_t start_idx, uint32_t end_idx, FunctionType func) const {
```
Comment (geirst) [inline vespalib/src/vespa/vespalib/btree/btreenode.h:374]: Please add doc describing function.

**Review by geirst — APPROVED** [review]
👍  Please add some more comments

File: vespalib/src/vespa/vespalib/btree/btreeiterator.h (hunk 1 above)
Comment (toregge) [reply vespalib/src/vespa/vespalib/btree/btreeiterator.h:313]: Done.

File: vespalib/src/vespa/vespalib/btree/btreeiterator.h (hunk 3 above)
Comment (toregge) [reply vespalib/src/vespa/vespalib/btree/btreeiterator.h:522]: Done

File: vespalib/src/vespa/vespalib/btree/btreeiterator.h (hunk 2 above)
Comment (toregge) [reply vespalib/src/vespa/vespalib/btree/btreeiterator.h:334]: Done.

File: vespalib/src/vespa/vespalib/btree/btreeiterator.h (hunk 4 above)
Comment (toregge) [reply vespalib/src/vespa/vespalib/btree/btreeiterator.h:538]: Done.

File: vespalib/src/vespa/vespalib/btree/btreeiterator.h (hunk 5 above)
Comment (toregge) [reply vespalib/src/vespa/vespalib/btree/btreeiterator.h:503]: Done.

File: vespalib/src/vespa/vespalib/btree/btreenode.h (hunk 6 above)
Comment (toregge) [reply vespalib/src/vespa/vespalib/btree/btreenode.h:486]: Done.

File: vespalib/src/vespa/vespalib/btree/btreenode.h (hunk 7 above)
Comment (toregge) [reply vespalib/src/vespa/vespalib/btree/btreenode.h:378]: Done.

**Review by geirst — APPROVED** [review]
👍

**Pull request merged**